### DISCUSSION TITLE
How do you get consistent snapshot testing across environments?

### DISCUSSION THREAD

**Discussion by RVledder** [cypress-io/cypress#24688] [Questions and Help]
In our project, we're using snapshot testing. However, some devs are running Windows, some are running OSX and the CI is using Linux. Rendering the snapshots could be different on either one of the environments. (fonts, shadow rendering)

How do you solve this issue?

What we're currently doing is running Docker with Linux locally to match the CI. But it feels a bit odd to be having to use Linux, a system that our user base doesn't use. Also we're currently experiencing an issue with running Docker locally, which raises this question, why use Docker at all for this.

**Comment by portyu9** [comment]
Hi @RVledder 

Using the same canonical rendering environment for the visual-regression baseline is the right approach. For pixel-level snapshot tests, comparing screenshots generated on Windows/macOS against a Linux baseline will inherently produce noise from fonts, anti-aliasing, shadows, browser versions, display scaling, etc.

So using Docker locally to reproduce the same environment as CI is not unusual. I would make the visual-test environment deterministic:

- use the same OS/container image
- pin the browser version
- use a fixed viewport
- install the same fonts
- disable or stabilize animations/transitions

The visual-regression environment does **not** have to represent the OS most of your users run. Its purpose is to detect unintended visual changes consistently.

If you specifically need platform-specific rendering coverage, keep separate baselines for each OS/browser combination instead of comparing all platforms against one baseline.

You can still run normal functional Cypress tests across the browsers/platforms your users actually use; just keep pixel-diff testing on a controlled rendering environment.

Please mark this answer as accepted if it helped, thank you.
